doc/man/ejsvm.1
.TH EJSVM "1" "March 2009" "ejsvm" "User Commands"
.SH NAME
ejsvm \- Ejscript Virtual Machine.
.SH SYNOPSIS
.B ejsvm
[\fI--class className\fR] 
[\fI--modules "modules..."\fR]
[\fI--log logSpec\fR]
[\fI--method methodName\fR]
[\fI--searchPath ejsPath\fR]
[\fI--version\fR]
module [arguments ...]
.SH DESCRIPTION
The \fBejsvm\fR command runs Ejscript programs by loading and executing modules previously compiled by 
the \fBejsc\fR Ejscript compiler.  The \fBejsvm\fR command provides a Virtual Machine environment for ejscript 
programs complete with class loader, byte code execution, garbage collection and core system types.
.PP
If you wish to run Ejscript programs interactively or to parse and execute in one-step, use the \fBejs\fR command shell.
.SH INVOCATION
.PP
The \fBejsvm\fR command can optionally invoke a specific static method on startup. If either the --class or --method switches are 
supplied, then \fBejsvm\fR will run the requested static method after the input modules have been loaded and all 
global scripts have been run.
.SH ORDERING
The \fBejsvm\fR command will load the specified modules and all dependent modules. Dependent modules
are specified in programs via the \fBuse module\fR directive. After that \fBejsvm\fR will load any 
additional modules specified via the \fB--modules\fR switch including with their dependent
modules. After each module and its dependent modules are loaded, the global statements of the module will be executed.
.SH OPTIONS
.TP
\fB\--class className\fR
Use the given className when invoking the method specified via --method.
.TP
\fB\--module "module ..."\fR
Normally modules are loaded automatically by the Ejscript loader in response to \fBuse module\fR directives in 
scripts, but you can use the \fB--modules\fR switch to load modules that contain global script code outside
any module directive.
.TP
\fB\--log logName[:logLevel]\fR
Specify a file to log internal execution messages. Ejs will log execution related trace to the log file. The log level
specifies the desired verbosity of output. Level 0 is the least verbose and level 9 is the most. Level 2 will include
trace of each source statement and assembler instruction executed (provided the source was compiled with --debug).
.TP
\fB\---method methodName\fR
Set the startup method name. Defaults to \fBmain\fR if a --className is specified and --method is not.
.TP
\fB\--searchPath ejsPath\fR
Override the module search path. The module search path is a set of directories that the \fBejsvm\fR command will use
when locating and loading Ejscript modules. Given a module named "a.b.c" in a script, \fBejsvm\fR will use the following 
search strategy to locate the module:
.IP
1. Search for a module file named "a.b.c.mod"
.IP
2. Search for a module file named "a/b/c.mod"
.IP
3. Search for a module file named "a.b.c.mod" in EJSPATH
.IP
4. Search for a module file named c.mod in EJSPATH
.IP
The search path is initially specified via the environment variable EJSPATH and may be overridden via the --searchPath ejsPath
switch. EJSPATH and the ejsPath command line value are similar to the system PATH formats. On windows, path segments 
are separated by ";" and on Linux, Unix, FreeBSD and MAC, the path segments are separated by ":" delimiters.
.TP
\fB\--version\fR
Print the \fBejsvm\fR command version and exit.
.PP
.SH BACKGROUND
Ejscript is an enhanced implementation of the JavaScript language for use in embedded applications such as web 
servers, embedded and mobile devices. It is especially well suited for Server-Side JavaScript web applications.
.PP
Ejscript is a dynamic, interpreted, object-oriented scripting language that supports classes, objects, exceptions, 
statements, expressions and a powerful suite of data types.
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2009 Embedthis Software.  Ejscript is a trademark of Embedthis Software.
.br
.SH "SEE ALSO"
ejsc, ejsgen, ejs, ejsmod


src/cmd/ejsvm.c
/**
 *  ejsvm.c - C Language VM to interpret Ejscript programs.
 *
 *  Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejs.h"
 
/****************************** Forward Declarations **************************/

static int  loadModuleFile(Ejs *ejs, char *fileName);

/************************************ Main ************************************/

int main(int argc, char *argv[]) 
{
    Mpr             *mpr;
    EjsService      *ejsService;
    Ejs             *ejs; 
    cchar           *programName;
    char            *argp, *className, *methodName, **modules, *extraModules, *module;
    int             i, flags, nextArg, err, count;

    err = 0;
    className = 0;
    methodName = 0;
    extraModules = 0;
    flags = 0;

    /*
     *  Create the Embedthis Portable Runtime (MPR) and setup a memory failure handler
     */
    mpr = mprCreate(argc, argv, ejsMemoryFailure);
    mprSetAppName(mpr, argv[0], 0, 0);

    if (mprStart(mpr, 0) < 0) {
        mprError(mpr, "Can't start mpr services");
        return EJS_ERR;
    }
    
    programName = mprGetBaseName(argv[0]);

    ejsService = ejsCreateService(mpr);
    if (ejsService == 0) {
        mprError(mpr, "Can't initialize the Ejscript service.");
        return -1;
    }

    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (strcmp(argp, "--class") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                className = argv[++nextArg];
            }

        } else if (strcmp(argp, "--log") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                ejsStartLogging(mpr, argv[++nextArg]);
            }

        } else if (strcmp(argp, "--method") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                methodName = argv[++nextArg];
            }

        } else if (strcmp(argp, "--modules") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                extraModules = argv[++nextArg];
            }

        } else if (strcmp(argp, "--searchpath") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                ejsSetSearchPath(ejsService, argv[++nextArg]);
            }

        } else if (strcmp(argp, "--version") == 0 || strcmp(argp, "-V") == 0) {
            mprErrorPrintf(mpr, "%s %s\n"
                "Copyright (C) Embedthis Software 2003-2009\n"
                "Copyright (C) Michael O'Brien 2003-2009\n",
               BLD_NAME, BLD_VERSION);  
            exit(0);

        } else {
            err++;
            break;
        }
    }

    if (nextArg == argc) {
        err++;
    }

    if (err) {
        mprErrorPrintf(mpr, 
            "Usage: %s [options] [module args] ...\n"
            "  Ejscript virtual machine options:\n"
            "  --class className         # Name of class containing method to run\n"
            "                            # Default: first class with main method\n"
            "  --modules \"modules...\"    # Extra modules to load\n"
            "  --log logSpec             # Internal VM diagnostics logging\n"
            "  --method methodName       # Name of method to run. Default: main\n"
            "  --searchpath ejsPath      # Module search path\n"
            "  --version                 # Emit the compiler version information\n\n", programName);
        return -1;
    }

    ejs = ejsCreate(ejsService, NULL, flags);
    if (ejs == 0) {
        mprError(mpr, "Can't create the EJSCript interpreter");
        mprFree(mpr);
        exit(-1);
    }
    
    if (extraModules) {
        mprMakeArgv(ejs, 0, extraModules, &count, &modules);
        for (i = 0; i < count && !err; i++) {
            if (loadModuleFile(ejs, modules[i]) < 0) {
                err++;
            }
        }
    }

    /*
     *  Load the module supplied on the command line. 
     */
    module = argv[nextArg];

    //  TODO - should have an API for this
    ejs->argc = argc - nextArg;
    ejs->argv = &argv[nextArg];

    if (loadModuleFile(ejs, module) < 0) {
        err++;
    }

    if (!err) {
        if (ejsRunProgram(ejs, className, methodName) < 0) {
            err++;
        }
    }

    mprFree(mpr);
    return err;
}



/*
 *  Load a module file
 */
static int loadModuleFile(Ejs *ejs, char *fileName)
{
    int         len;

    /*
     *  Strip trailing ".mod" if present
     */
    fileName = mprStrdup(ejs, fileName);
    len = strlen(fileName);
    if (len > 4 && strcmp(&fileName[len - 4], EJS_MODULE_EXT) == 0) {
        fileName[len - 4] = '\0';
    }

    if (ejsLoadModule(ejs, fileName, NULL, NULL, 0) == 0) {
        ejsReportError(ejs, "Can't load module \"%s\"", fileName);
        mprFree(fileName);
        return EJS_ERR;
    }

    mprFree(fileName);
    return 0;
}


/*
 *  @copy   default
 *  
 *  Copyright (c) Embedthis Software LLC, 2003-2009. All Rights Reserved.
 *  Copyright (c) Michael O'Brien, 1993-2009. All Rights Reserved.
 *  
 *  This software is distributed under commercial and open source licenses.
 *  You may use the GPL open source license described below or you may acquire 
 *  a commercial license from Embedthis Software. You agree to be fully bound 
 *  by the terms of either license. Consult the LICENSE.TXT distributed with 
 *  this software for full details.
 *  
 *  This software is open source; you can redistribute it and/or modify it 
 *  under the terms of the GNU General Public License as published by the 
 *  Free Software Foundation; either version 2 of the License, or (at your 
 *  option) any later version. See the GNU General Public License for more 
 *  details at: http://www.embedthis.com/downloads/gplLicense.html
 *  
 *  This program is distributed WITHOUT ANY WARRANTY; without even the 
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
 *  
 *  This GPL license does NOT permit incorporating this software into 
 *  proprietary programs. If you are unable to comply with the GPL, you must
 *  acquire a commercial license to use this software. Commercial licenses 
 *  for this software and support services are available from Embedthis 
 *  Software at http://www.embedthis.com 
 *
 *  @end
 */
